huaweicloud/services/iotda/resource_huaweicloud_iotda_device_certificate.go

```diff
@@ -0,0 +1,212 @@
+	}
+
+	d.SetId(*resp.CertificateId)
+	return resourceDeviceCertificateRead(ctx, d, meta)
+}
+
+func resourceDeviceCertificateRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	c := meta.(*config.Config)
+	region := c.GetRegion(d)
+	client, err := c.HcIoTdaV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	detail, err := QueryDeviceCertificate(client, d.Id(), utils.StringIgnoreEmpty(d.Get("space_id").(string)))
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "error retrieving IoTDA device CA certificate")
+	}
+
+	status := "Unverified"
+	if detail.Status != nil && *detail.Status {
+		status = "Verified"
+	}
+
+	mErr := multierror.Append(
+		d.Set("region", region),
+		d.Set("cn", detail.CnName),
+		d.Set("owner", detail.Owner),
+		d.Set("status", status),
+		d.Set("verify_code", detail.VerifyCode),
+		d.Set("effective_date", detail.EffectiveDate),
+		d.Set("expiry_date", detail.ExpiryDate),
+	)
+
+	return diag.FromErr(mErr.ErrorOrNil())
+}
+
+func resourceDeviceCertificateUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	c := meta.(*config.Config)
+	region := c.GetRegion(d)
+	client, err := c.HcIoTdaV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	opts := model.CheckCertificateRequest{
+		CertificateId: d.Id(),
+		ActionId:      "verify",
+		Body: &model.VerifyCertificateDto{
+			VerifyContent: d.Get("verify_content").(string),
+		},
+	}
+
+	_, err = client.CheckCertificate(&opts)
+	if err != nil {
+		return diag.Errorf("error verifing IoTDA device CA certificate: %s", err)
+	}
+
+	return resourceDeviceCertificateRead(ctx, d, meta)
+}
+
+func resourceDeviceCertificateDelete(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	c := meta.(*config.Config)
+	region := c.GetRegion(d)
+	client, err := c.HcIoTdaV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	deleteOpts := &model.DeleteCertificateRequest{
+		CertificateId: d.Id(),
+	}
+	_, err = client.DeleteCertificate(deleteOpts)
+	if err != nil {
+		return diag.Errorf("error deleting IoTDA device CA certificate: %s", err)
+	}
+
+	return nil
+}
+
+func QueryDeviceCertificate(client *v5.IoTDAClient, id string, spaceId *string) (*model.CertificatesRspDto, error) {
+	var marker *string
+	for {
+		resp, err := client.ListCertificates(&model.ListCertificatesRequest{
+			AppId:  spaceId,
+			Limit:  utils.Int32(50),
+			Marker: marker,
+		})
+
+		if err != nil {
+			return nil, err
+		}
+		if resp.Certificates == nil || len(*resp.Certificates) == 0 {
+			break
+		}
+
+		for _, v := range *resp.Certificates {
+			if utils.StringValue(v.CertificateId) == id {
+				return &v, nil
+			}
+		}
+		marker = resp.Page.Marker
+	}
+
+	return nil, golangsdk.ErrDefault404{}
```

Comment:
suggest to use github.com/huaweicloud/huaweicloud-sdk-go-v3/core/sdkerr.ServiceResponseError

Reply:
ok